src/smt/optimization_solver.h

```diff
@@ -80,38 +90,140 @@ class Objective
-    Objective d_activatedObjective;
-    /** A saved value of the objective from the last sat call. **/
-    Node d_savedValue;
-  };
+
+  /** Specify whether to use signed or unsigned comparison
+   * for BitVectors (only for BitVectors), this variable is defaulted to true
+   * **/
+  bool d_isSigned;
+};
+
+
+/**
+ * A solver for optimization queries.
+ *
+ * This class is responsible for responding to optmization queries. It
+ * spawns a subsolver SmtEngine that captures the parent assertions and
+ * implements a linear optimization loop. Supports activateObjective,
+ * checkOpt, and objectiveGetValue in that order.
+ */
+class OptimizationSolver
+{
+ public:
+  /** parent is the smt_solver that the user added their assertions to **/
+  OptimizationSolver(SmtEngine* parent);
+  ~OptimizationSolver();
+
+  /** Runs the optimization loop for the activated objective **/
+  OptResult checkOpt();
+  /** Activates an objective: will be optimized for
+   * Parameter is_signed specifies whether we should use signed/unsigned
+   * comparison for BitVectors (only effective for BitVectors) **/
+  void activateObj(const Node& obj,
+                   const int& type,
+                   bool bvSignedCompare = true);
+  /** Gets the value of the optimized objective after checkopt is called **/
+  Node objectiveGetValue();
+
+ private:
+  /** The parent SMT engine **/
+  SmtEngine* d_parent;
+  /** The objectives to optimize for **/
+  Objective d_activatedObjective;
+  /** A saved value of the objective from the last sat call. **/
+  Node d_savedValue;
+};
+
+/**
+ * Optimizer for individual datatype 
+ * Currently supported: Integer, BitVector
+ */
+struct OMTOptimizer 
+{
+  virtual ~OMTOptimizer() = default;
+  /** Given a target node, retrieve an optimizer specific for the node's datatype 
+   * the second field isSigned specifies whether we should use signed comparison for BitVectors 
+   * and it's only valid when the datatype is BitVector
+   **/
+  static std::unique_ptr<OMTOptimizer> getOptimizerForNode(CVC4::Node targetNode, bool isSigned=false);
+  /** Minimize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) = 0;
+  /** Maximize the target node with constraints encoded in parentSMTSolver 
+   * Parameters: 
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints 
+   * - target: the target expression to optimize 
+   * Return value: 
+   * - std::pair<OptResult, CVC4::Node>: the result of optimization and the optimized value 
+   **/
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) = 0;
```

Comment:
What's wrong with using ObjectiveType as before as a parameter rather than having separate functions for maximize and minimize?

Reply:
this is the optimizer, and fusing minimize and maximize together makes it messy.

Reply:
similarly to the return value issue above, this was based on the API. Perhaps would you like to check optimathsat and z3's optimization api for this as well?

Reply:
nope, this is coding style. The API is not defined here. This is just a private class that is called by the OptimizationSolver

Reply:
so the OptimizationSolver will check the target's type, if it's Integer then creates an OMTOptimizerInt and call OMTOptimizerInt::minimize/maximize. However OMTOptimizer is designed does not affect the outer API.

Reply:
I would stop discussing this. this is only an aesthetic perspective.